Add address lookup page

gclssvglx opened the pull request:
This completes the address lookup feature.

Adds a new page which handles looking up an addresses for a given postcode, selecting a specific address and confirming an address.

Also, stores a Unique Property Reference Number (UPRN) for the selected address.

We want to provide an address lookup to ensure that we collect correct addresses from users, and UPRN numbers where possible.

[Trello](https://trello.com/c/WoKUCfci/376-address-lookup-page)

Co-authored-by: Jon Hallam <jonathan.hallam@digital.cabinet-office.gov.uk>

gclssvglx committed:
Add address lookup page

This completes the address lookup feature.

Adds a new page which handles looking up an addresses for a given
postcode, selecting a specific address and confirming an address.

Also, stores a Unique Property Reference Number (UPRN) for the
selected address.

We want to provide an address lookup to ensure that we collect
correct addresses from users, and UPRN numbers where possible.

Co-authored-by: Jon Hallam <jonathan.hallam@digital.cabinet-office.gov.uk>

bilbof commented:
Rebased 👍

bilbof commented on app/helpers/address_helper.rb:
```diff
@@ -9,7 +9,7 @@ def postcode_lookup(postcode)
-    raise AddressLookupError, details["error"] unless response.status == 200
+    raise AddressLookupError, response.status unless response.status == 200
 
     addresses = details["results"]
```
`addresses` can be `nil`, so calling `addresses.map` below will fail. You could do `details.fetch("results", [])` and also check if `"totalresults"` is 0.

JonathanHallam replied:
Heya, can you explain this a bit more. As far as we knew this was taking all of the addresses for a given postcode. Does this mean that there are postcodes without addresses? Do you have and example? 

We were expecting this to either return an empty array or to return a 400 before it got to this point.

gclssvglx replied:
Updated to use response.status.  Yes, there are postcodes without addresses (e.g. SW1A3AA) that we have accounted for.

bilbof reviewed (commented):
There doesn't seem to be a way to edit a postcode after you enter an incorrect one. We should probably make that possible.

bilbof reviewed (dismissed):
We should the keys such as `POSTCODE_LOCATOR` (postcode) rather than parsing the raw `ADDRESS` field if possible.

E.g. if you enter the FCO canteen postcode "sw1a2ae":

<img width="731" alt="Screenshot 2020-05-14 at 14 07 35" src="https://user-images.githubusercontent.com/8124374/81938073-550da700-95ec-11ea-9113-d7ca0536bbe7.png">

bilbof commented:
We should probably get some error messages approved by our content people.

<img width="685" alt="Screenshot 2020-05-14 at 14 05 31" src="https://user-images.githubusercontent.com/8124374/81937821-03651c80-95ec-11ea-9d1e-ad66bb33fdc1.png">

bilbof commented on app/controllers/coronavirus_form/address_lookup_controller.rb:
```diff
@@ -0,0 +1,54 @@
+
+class CoronavirusForm::AddressLookupController < ApplicationController
+  def submit
+    redirect_to check_your_answers_url if session[:check_answers_seen]
```
It looks like this would prevent people from looking up a new address. Is that something we want to permit?

bilbof replied:
Additionally, if someone goes back to this page, and re-enters a postcode, they get a 500 error, due to a DoubleRenderError (redirect_to and then a render controller_path called later).

JonathanHallam replied:
Now people can look up a new address but they have to go back to do it. They can either use the back links or browser buttons. Additionally the extra page we added seems to have fixed the 500 error

bilbof commented on app/controllers/coronavirus_form/address_lookup_controller.rb:
```diff
@@ -0,0 +1,54 @@
+private
+
+  def update_params
+    uprn_and_address = params[:address_string].split("|")
```
If you use JSON in the input values you could then parse that as JSON here, rather than parsing a string.

E.g. 

```html
<select>
   <option value='{"uprn":"1234","postcode":"abc","address_line_1":"10 Downi..."..}'>
      10 Downing Street, London..
    </option>
</select>
```

```ruby
address = validate_address(JSON.parse(selected_address))
```

This should make it easier to parse and validate the selected address.

gclssvglx replied:
In order to make this work we had to make another API call to the uprn end-point.  We are not using JSON, but we are using all of the fields directly from the API response, so strings are no longer a problem.

bilbof commented on app/controllers/coronavirus_form/address_lookup_controller.rb:
```diff
@@ -0,0 +1,54 @@
+      redirect_to support_address_path
+    else
+      begin
+        results = helpers.postcode_lookup(params[:postcode])
```
We should probably validate the postcode here using our postcode validator. No point making an API call with a bad postcode.

gclssvglx replied:
Done.

bilbof commented on app/controllers/coronavirus_form/address_lookup_controller.rb:
```diff
@@ -0,0 +1,54 @@
+        value: "#{address.first.first} | #{address.first.last}",
+      }
+    end
+  end
```
I think it'd be better if the API / helper class you have could provide the data in a format the controller requires, rather than putting extra logic into the controller. Ideally controllers are dumb and don't do anything other than call other classes

gclssvglx replied:
Done.  We know use the fields from the API response directly.

bilbof commented on app/controllers/coronavirus_form/support_address_controller.rb:
```diff
@@ -7,15 +7,25 @@ class CoronavirusForm::SupportAddressController < ApplicationController
-        county: strip_tags(params[:county]&.strip).presence,
-        postcode: strip_tags(params[:postcode]&.gsub(/[[:space:]]+/, "")).presence,
-      },
+      support_address: helpers.compare_and_return(session[:support_address].symbolize_keys, given_address),
```
I'm not really clear on the intent of the `compare_and_return` method.

gclssvglx replied:
Fixed! See code.

bilbof commented on app/errors/address_lookup_error.rb:
```diff
@@ -1,5 +1,5 @@
-  def initialize(message)
-    super(message)
+  def initialize(code)
+    super(I18n.t("coronavirus_form.api_errors.#{code}"))
```
It would be worth having a default here, since it could be that you receive a different status code, or `nil` as the code.

JonathanHallam replied:
Have added in a default value :)

bilbof commented on config/locales/en.yml:
```diff
@@ -233,6 +233,11 @@ en:
+      400: "Bad Request | E.g. missing query parameter, malformed syntax."
+      401: "Unauthorized | The client has not provided authentication or incorrect authentication."
+      404: "Not Found | The server has not found anything matching the Request-URI."
+      500: "Internal Server Error| Generic internal server error."
```
```suggestion
      500: "Internal Server Error | Generic internal server error."
```

However, I think a content person / service designer will likely want to write these.

joelanman replied:
https://design-system.service.gov.uk/patterns/service-unavailable-pages/

gclssvglx replied:
Have made changes to the copy and no longer use these responses from OS.

bilbof commented on spec/support/fill_in_the_form_steps.rb:
```diff
@@ -65,6 +65,25 @@ def and_has_given_their_date_of_birth
+    end
+    VCR.turn_off!
+  end
+
```
I think it would be useful to have tests for missing addresses and other errors here.

gclssvglx replied:
Have added to other tests.

bilbof commented on spec/controllers/coronavirus_form/address_lookup_controller_spec.rb:
```diff
@@ -0,0 +1,62 @@
+          expect(response).to render_template(current_template)
+        end
+      end
+    end
```
Would be good to have a test for what happens if they have already viewed the check_your_answers page here.

gclssvglx replied:
Done.

bilbof commented:
Great work so far 👍 I think there are a few changes that need to be made to the code, and to get designers / content folks to have a good look over this too :)

JonathanHallam commented:
> There doesn't seem to be a way to edit a postcode after you enter an incorrect one. We should probably make that possible.

We think we're going to have to make another route in order to do this. Will make a start on that on Monday

bilbof commented:
Another thing that would be useful is if this could still work locally / in Heroku... I wonder if we could redirect people to the "enter address manually" page if - for whatever reason - the API isn't working (e.g. it goes down / the api key isn't working)

gclssvglx committed:
Handle 401 - Invalid Key

In the event we get a 401 from the OS Places API we know that our key has become invalidated.  There is nothing the user can do about this, so we redirect them to input their address manually.  We also log an error with Sentry.

Co-authored-by: Jon Hallam <jonathan.hallam@digital.cabinet-office.gov.uk>

bilbof commented on app/controllers/coronavirus_form/address_lookup_controller.rb:
```diff
@@ -0,0 +1,33 @@
+  rescue AddressLookupError, AddressNotFoundError => e
+    session[:flash] = e.message
+
+    if e.class.name == "AddressLookupError" && e.status == 401
```
I would recommend raising a different error in the helper for the authentication error (e.g. `AddressAuthError`), then you can have a simpler controller, e.g.

```ruby
rescue AddressAuthError
      GovukError.notify("API Key is invalid!")
      redirect_to support_address_path
rescue AddressLookupError, AddressNotFoundError => e
      session[:flash] = e.message
      redirect_to postcode_lookup_url
```

JonathanHallam replied:
Nice shout, have just pushed this change up :)

bilbof commented on app/controllers/coronavirus_form/address_lookup_controller.rb:
```diff
@@ -0,0 +1,33 @@
+
+  def submit
+    if session[:check_answers_seen]
+      redirect_to check_your_answers_url
```
Do we want to save the new address if the user submit on this page again?

gclssvglx replied:
See comment on `postcode_lookup_controller`.

bilbof commented on app/controllers/coronavirus_form/address_lookup_controller.rb:
```diff
@@ -0,0 +1,33 @@
+    if session[:check_answers_seen]
+      redirect_to check_your_answers_url
+    else
+      address = helpers.uprn_lookup(params[:uprn])["results"][0]["LPI"]
```
You could use `dig` here, I.e. `helpers.uprn_lookup(uprn).dig('results',0,'LPI')`, however, I wonder if this logic (including `convert_address(address)`) needs to happen in the controller? Could the helper just return the already converted address?

JonathanHallam replied:
Nice spot on the dig, we've adjusted. On the 2nd point, we'd rather leave these as 2 separate calls. This makes it easier to test because it avoids having to chain methods together in the helper. We could maybe avoid this by having a third method along the lines of `lookup_address_by_uprn_and_convert` but I really don't like that.

bilbof commented on app/controllers/coronavirus_form/postcode_lookup_controller.rb:
```diff
@@ -0,0 +1,39 @@
+
+class CoronavirusForm::PostcodeLookupController < ApplicationController
+  def show
+    if session[:flash].present?
```
I wonder if you could avoid setting error messages and the session data by using the same controller for both the address picker and the postcode lookup? You'd have different views / endpoints, but you could store the errors etc in the controller instance variables and wouldn't need to do redirects if lookups failed, you'd just render a different view.

You wouldn't need to add and then remove values from the session, like `session[:postcode]`, since this could all be set in controller instance variables.

JonathanHallam replied:
When you first reviewed this we did have everything in one controller but that was mainly because there was only one route. In order to improve the format in which full addresses we displayed after selection we felt that we needed the extra route. Going back to a single controller way of doing things would help solve the issues that you mention above but would make that controller quite heavy with logic which we'd be keen to avoid. 

You'd also still be in the position where you'd have to redirect to the support address path in the event you get a 401 out of the OS API. We think it's best to show that error message on the support-address page so even if we were to do this at the very least we'd need to be storing that error message in the session.

Storing the error message in the session is a bit of an anti pattern but having the two controllers does cut down on controller logic and avoid us having a controller that was several hundred lines long. Happy to talk about this offline.

bilbof commented on app/controllers/coronavirus_form/postcode_lookup_controller.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  def submit
+    redirect_to check_your_answers_url if session[:check_answers_seen]
```
This'll prevent people from looking up another postcode, even if they enter a new one. Is that behaviour desired?

gclssvglx replied:
Removed the redirect from here as it was possible to enter a new address without saving it ini the session.

bilbof commented on app/controllers/coronavirus_form/postcode_lookup_controller.rb:
```diff
@@ -0,0 +1,39 @@
+
+    if params[:postcode].blank?
+      flash.now[:validation] = [{ text: I18n.t("coronavirus_form.questions.address_lookup.hint"), field: "postcode" }]
+      render controller_path, status: :unprocessable_entity
```
If I don't enter an address, the error message is "Select your address", when it should be "Enter your postcode". Is that intended?

JonathanHallam replied:
Ooops, absolutely not. Have fixed, good spot 🥇

bilbof commented on app/controllers/coronavirus_form/postcode_lookup_controller.rb:
```diff
@@ -0,0 +1,39 @@
+      flash.now[:validation] = [{ text: I18n.t("coronavirus_form.questions.address_lookup.hint"), field: "postcode" }]
+      render controller_path, status: :unprocessable_entity
+    else
+      errors = helpers.validate_postcode("postcode", params[:postcode])
```
I don't think you need to use `helpers` here, you can just call `validate_postcode`

bilbof replied:
I wonder also whether `params[:postcode].blank?` could be done by the `validate_postcode` method? Then you could have less code in the controller (e.g. only one `render controller_path, status: :unprocessable_entity`), which is usually a good thing.

JonathanHallam replied:
We tried removing the helpers calls here and it caused a few method errors in the tests.

Just had a look at this and yes it does seem a bit unnecessary to have that call here but the `validate_postcode` method already does a check for a blank postcode. It does this to stop itself falling over when it gets to the regex that checks format. It is possible to input an error if postcode is blank in the `validate_postcode` method. If we want to move this check to the `validate_postcode` method we have two options. 

One is to remove `POSTCODE` from the `REQUIRED_FIELDS` constant in the `support_address_controller`. This will stop the validation happening twice but is quite horrible because we'd kind of be saying that postcode both was and wasn't mandatory. 

The other would be to move the `validate_missing_fields` method from the `support_address_controller` into the `field_validation_helper`. This is doable but would mean moving quite a bit of stuff around. It would also make the i18n structure a bit brittle because we'd want it to be universal for all of the required fields across the form. 

I think this would be worth more thought if we started doing this in more cases. As it stands though, postcode is a bit of a snowflake because of the regex validation that we're doing in the `field_validation_helper`. Because of this I don't think it's worth the downsides / effort.

bilbof commented on app/controllers/coronavirus_form/address_lookup_controller.rb:
```diff
@@ -0,0 +1,33 @@
+
+    if e.class.name == "AddressLookupError" && e.status == 401
+      GovukError.notify("API Key is invalid!")
+      redirect_to support_address_path
```
It might be a bit confusing for users if this happens. Should we show them an error message?

JonathanHallam replied:
We do. We figured that if OS was returning a 401 it meant that someone had revoked our key and that there was nothing the user could do. In the event that this happens it redirects to a blank support-address form with a message that says `An error has occurred, please enter your address manually`

You can test this locally by replacing your api key with nonsense and looking up an address.

bilbof commented on app/controllers/coronavirus_form/support_address_controller.rb:
```diff
@@ -1,53 +1,75 @@
-      redirect_to check_your_answers_url
     else
+      @form_responses = {
+        support_address: helpers.update_address(session[:support_address].symbolize_keys, given_address),
```
Here could you set `@form_responses = helpers.update_address ...` rather than it being a hash?

JonathanHallam replied:
nice, this clears it up some, thanks :)

bilbof commented on app/errors/address_lookup_error.rb:
```diff
@@ -1,5 +1,8 @@
+  def initialize(status)
+    @status = status
+    @status = "error" unless I18n.t("coronavirus_form.api_errors").keys.include?(@status)
+    super(I18n.t("coronavirus_form.api_errors.#{@status}"))
```
You could use a `default` here instead of the `unless` line: https://guides.rubyonrails.org/i18n.html#defaults

e.g. `I18n.t("coronavirus_form.api_errors.#{@status}", default: "coronavirus_form.api_errors.error")`

JonathanHallam replied:
didn't know this was a thing, nice!

bilbof commented on app/helpers/address_helper.rb:
```diff
@@ -1,19 +1,104 @@
-  API_URL = "https://api.ordnancesurvey.co.uk/places/v1/addresses/postcode?dataset=LPI"
+  UPRN = %w[UPRN].freeze
+  SAO = %w[ORGANISATION SAO_START_NUMBER SAO_START_SUFFIX SAO_END_NUMBER SAO_END_SUFFIX SAO_TEXT].freeze
+  PAO = %w[PAO_START_NUMBER PAO_START_SUFFIX PAO_END_NUMBER PAO_END_SUFFIX PAO_TEXT].freeze
```
Oh my word, this looks hard! You must know so much about addresses now 👏

JonathanHallam replied:
yeah, more than we'd like! If you want talking through this then we'd be happy to. Also happy to do a brief show and tell to the other devs on how addresses work once this story's in if you think that'd be useful :)

bilbof replied:
Definitely that would be useful :) I'm sure this would be something for the wider gov.uk show and tell, too.

bilbof reviewed (commented):
Great progress 👍 I've reviewed the controllers / logic, but wanted to give some immediate feedback, before re-reviewing.

I found the postcode lookup request quite slow (maybe a few seconds). Is there a way to do just one request to the OS API? I wasn't sure why we'd need to make two requests there.

JonathanHallam commented:
> Great progress 👍 I've reviewed the controllers / logic, but wanted to give some immediate feedback, before re-reviewing.
> 
> I found the postcode lookup request quite slow (maybe a few seconds). Is there a way to do just one request to the OS API? I wasn't sure why we'd need to make two requests there.

@bilbof 

So we could do it with one request but that would mean that we'd either have to rely heavily on JS or be saving a potentially massive hash to the session. The problem is that the user has to choose their address from the list. This kind of makes it necessary to have an extra route which, along with wanting to use the json fields to put chosen address data into the support-address page is what necessitates the second call. 

It's also worth noting that the initial call to the api after you input a postcode would take exactly the same amount of time. The extra time spent by the 2nd call is only what's spent between clicking continue after choosing an address from the dropdown and seeing that address on the support-address page.

gclssvglx committed:
Fixed bug with check answers page

We noticed that UPRN was showing up on the check answers page and felt that this is not useful to user so removed it from the view only.

The link to change postcode was incorrect in order to fix this we had to update the en.yml file to use the same key convention as other routes.

Removed redirect_to if check answers seen for the postcode and address lookup pages. This means that a user can no longer change their address without changing the address as stored in the session.  In short, if a user wants to change their address or postcode they have to go through the support_address page before being redirected to the check answers page.

Co-authored-by: Jon Hallam <jonathan.hallam@digital.cabinet-office.gov.uk>

Commit:
Remove second API call to OS API

This also includes minor refactorings/changes.

bilbof reviewed (commented):
I'm happy for this to be merged. Great work @JonathanHallam @gclssvglx! 👏 

It'd be good to have the product and design folks on the team give this a thorough review / QA with the Heroku app, so if you could share the running review app with them that'd be a good next step. I'll add a green tick here once they've approved.

I'd quite like some brief docs on how this works, not only so we can maintain this repo, but also so others can easily copy this pattern elsewhere. If you have ideas of how this could be put into a pattern/component, those docs would be a good place to put them.

gclssvglx committed:
Amend copy on postcode lookup page

This change makes it clear that the postcode the user is entering is for the person who needs support.

Co-authored-by: Jon Hallam <jonathan.hallam@digital.cabinet-office.gov.uk>

bilbof reviewed (approved):
🎉

joelanman reviewed (changes requested):
The current postcode validation on the service strips out any whitespace before validating - I think we should stick with that logic, as we know what postcode the user meant, and whitespace is hard for a user to 'fix'.

The address lookup pattern in the Design System plays back the postcode on the dropdown page so a user can check it, and has a link to change it (back to the previous page) if its not correct - it also has a shorter label for the dropdown, can we have 'Select your address'?

joelanman commented:
Sorry forgot one - if the user enters an unrecognised postcode, they just get a validation error and cannot continue. We should probably link to 'Enter your address manually' in case of some issue that means the lookup isn't working in 100% of cases

joelanman commented on app/views/coronavirus_form/address_lookup.html.erb:
```diff
@@ -0,0 +1,39 @@
+<%= render "govuk_publishing_components/components/button", text: t("coronavirus_form.submit_and_next"), margin_bottom: true %>
+<% end %>
+
+<p class="govuk-body">
+  <%= t("coronavirus_form.questions.address_lookup.copyright") %>
+</p>
```
can we remove this? Not sure why it's been added

JonathanHallam replied:
This was on the AC for the card. I'm guessing that it's required because we're using copyrightable data from Ordinance Survey.

StephenGill replied:
Do we have a link to the licence?

joelanman replied:
it was on the card, but no one seems to know why. It's unusual to have a line like this in a service

gclssvglx replied:
Removed.

gclssvglx committed:
Strip all spaces from postcode before validation

Co-authored-by: Jon Hallam <jonathan.hallam@digital.cabinet-office.gov.uk>

gclssvglx committed:
Follow design pattern

Ensure postcode and address selection follows the pattern available here: https://design-system.service.gov.uk/patterns/addresses/.

Minor content change as requested.

Update copyright text.

Co-authored-by: Jon Hallam <jonathan.hallam@digital.cabinet-office.gov.uk>

joelanman reviewed (approved):
Great work!

bilbof commented:
LGTM. If we have stakeholder approval lets merge this tomorrow morning :)

joelanman commented:
Sorry, looked into the copyright thing - looks like the correct line is:

>Contains Royal Mail data © Royal Mail copyright and database rights 2020

Can we make it use the small font, and space it away from the form - mockup:

![image](https://user-images.githubusercontent.com/1132904/83427328-78c45000-a428-11ea-9409-dcf960473039.png)

joelanman commented:
Looks great thanks
